### DISCUSSION TITLE
Is UNO not viable for OpenGL applications on Windows due to WinUI 3 restrictions?

### DISCUSSION THREAD

**Discussion by KarlHedlund** [unoplatform/uno#11034] [Q&A]
I'm coming from several WPF and Xamarin projects so I've been pretty keen on working with more XAML based GUI frameworks. Since I'm in need of targeting at least Windows and Linux, I want a shared OpenGL renderering solution for 3D. This has worked well in the past. Having to maintain both a DirectX and OpenGL solution isn't very attractive.

I've got OpenGL working in a UNO WinUI application but while working with the OpenGL code something stopped working. I then went to check out what had happened with programs like apitrace, RenderDoc and NVIDIA nSight but none of them could hook into the application. This is when I learned that there is a special way UWP/WinUI programs are started and that therefore, these kinds of programs can't hook into them to read all the driver calls. I don't know the details of this yet, just that it isn't possible.

Not being able to debug the graphics code is really a showstopper. It looks like the only solution is to use DirectX for WinUI and that is something we don't want to do.

I think from UNO's perspective, I'm not the only one who thinks it would be great to use a unified OpenGL solution for all targeted platforms.

Thoughts?

**Comment by jeromelaban** [comment]
Windows apps are not explicitly managed by Uno Platform. You will need to ask questions about it [here](https://github.com/microsoft/microsoft-ui-xaml) or [here](https://github.com/microsoft/WindowsAppSDK).

**Reply by KarlHedlund** [reply]
I was wondering if there were other developers trying out UNO that had come to the same conclusion and perhaps together we could apply some pressure on Microsoft who are historically not keen on promoting OpenGL usage. It's a long-shot but I wanted to see if more people are in my situation. Maybe together we could show that this is important.
